Enhance getInputProps to allow passing of non-overridden props

joshuatf opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Previously, adding an `onChange` handler or anything else that collided with `getInputProps` would require recalling `getInputProps.someFunction` inside the new handler.  This PR introduces a second argument to `getInputProps` which passes props to the control without overriding the original handlers where possible.

Some props will not override the original handler (e.g., if you pass on `onChange` it will be combined with the `Form` component's `onChange`) and some will override (e.g., `value` can not be combined and the consumer prop will override the original value).

This PR also uncovered numerous TS issues, so a `getCheckboxProps` was introduced into the `Form` component to omit invalid props (e.g., `value` for checkboxes).  More specific prop getters are probably in order (`getSelectProps`), but I think we can iterate on these as we go.

Finally, a `sanitize` method was added as part of the optional consumer input props.  This is a shortcut to sanitizing the input value and handling changes on blur.

**TL;DR:**

* `getInputProps` now takes a second argument to combine/override props where necessary
* `sanitize` method can be added to an input to handle sanitization on blur
* `getCheckboxProps` was introduced for props specific to checkboxes

Closes #34996 .

### How to test the changes in this Pull Request:

1. Navigate to the new product page `Products -> Add new (MVP)`
2. Check that styling for checkboxes is still correct
3. Add values to the price and dimension fields, using thousand commas and decimal separators (depending your locale)
4. Make sure they are sanitized on blur and persist correctly on save

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> run changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

joshuatf committed:
Allow additional props to be passed to the Form getInputProps method

joshuatf committed:
Remove getTextControlProps

joshuatf committed:
Pass additional shared props through getInputProps in shipping

joshuatf committed:
Simplify checkbox props

joshuatf committed:
Unwrap currency props

joshuatf committed:
Use onBlur event to sanitize prices

joshuatf committed:
Add changelog entry

joshuatf commented on plugins/woocommerce-admin/client/products/sections/pricing-section.tsx:
```diff
@@ -135,10 +132,8 @@ export const PricingSection: React.FC = () => {
-								regularPriceProps?.onChange( sanitizedValue );
-							} }
+							value={ formatCurrencyDisplayValue( regularPriceProps?.value, currencyConfig, formatAmount ) }
+							onBlur={ () => setValue( 'regular_price', sanitizePrice( regularPriceProps.value ) ) }
```
@octaedro I moved the sanitization to `onBlur`.  I think this is a little bit more conventional as I was worried about input being sanitized as it was typed could lead to odd behavior.

But please let me know if I broke anything here or if this needs to be sanitized as we type.

octaedro replied:
Sounds good to me! Good idea!

joshuatf commented on plugins/woocommerce-admin/client/products/sections/utils.ts:
```diff
@@ -1,110 +1,79 @@
+		? 'prefix'
+		: 'suffix';
+
 	return {
```
My goal was to unwrap some of these so that we can:
1. More easily follow where props are being passed
2. Reduce the knowledge of form handlers in this function
3. Reduce overall logic in this function

joshuatf reviewed (commented):
Just noting that we have more upcoming work (I also have an open PR) that will benefit from being able to pass in additional handlers without overriding the `Form`'s handlers.

joshuatf committed:
Add option to get checkbox props to form context helpers

joshuatf committed:
Update checkbox tracks handler naming and typing

joshuatf committed:
Fix up usage of getInputProps

joshuatf committed:
Add helper sanitize method

joshuatf committed:
Use sanitize helper method for product input fields

joshuatf committed:
Fix inventory input props after rebase

joshuatf commented:
@octaedro Thanks for the ping on rebasing this one in Slack!  After rebasing, a lot of TS issues cropped up that I think we were able to errantly bypass before.

I added a `getCheckboxProps` method to `Form` and also introduced a `sanitize` method as a shortcut for inputs.  I updated the description above to reflect all the changes.

@mdperez86 since this also affects a lot of your recent additions around dimensions, it would be great to get your 👀  on this as well.

joshuatf committed:
Fix shipping typo

joshuatf commented on packages/js/components/src/form/form.tsx:
```diff
@@ -268,35 +297,59 @@ function FormComponent< Values extends Record< string, any > >(
 		};
 	}
 
+	function getCheckboxProps< Value = Values[ keyof Values ] >(
```
Should this be `getCheckboxInputProps` or `getCheckboxControlProps` instead?  The reason I ask is we'll probably want to introduce a `getSelectProps` which sounds odd and could be interpreted the wrong way.  Also okay leaving it as is, but wanted to get some feedback.

octaedro replied:
I would lean towards `getCheckboxControlProps`. But I'm also ok leaving it as it is now.

joshuatf replied:
I think that makes sense too since we're really creating props specifically for our controls 👍

octaedro commented on plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:
```diff
@@ -78,13 +84,9 @@ export const ProductDetailsSection: React.FC = () => {
-								setSkuIfEmpty();
-								getInputProps( 'name' ).onBlur();
-							} }
+							{ ...getInputProps( 'name', {
+								onBlur: setSkuIfEmpty,
+							} ) }
```
❤️

octaedro reviewed (commented):
Nice job @joshuatf! This is testing well here and the code looks good.
There is a small visual error with the `Feature this product` checkbox, the rest looks great!

![Screen Shot 2022-10-19 at 09 55 35](https://user-images.githubusercontent.com/1314156/196716535-455fdfe6-6b0d-43c6-9014-0c4311049026.png)

joshuatf committed:
Fix up form types after rebase

joshuatf committed:
Align all checkboxes on product page

joshuatf committed:
Rename new checkbox helper to getCheckboxControlProps

joshuatf commented:
>There is a small visual error with the Feature this product checkbox, the rest looks great!

Good catch @octaedro!  Updated in https://github.com/woocommerce/woocommerce/pull/35034/commits/38f80d8cca8ec53ef5b98000b6cb82fe8db64b29

Also gave this a rebase; could you take a look when you have time?

joshuatf committed:
Add helper method to get select control props

joshuatf committed:
Add data changelog entry

joshuatf committed:
Check for product name length on blur

joshuatf committed:
Add initial value for name to prevent uncontrolled value

joshuatf committed:
Add initial value for sku

octaedro reviewed (dismissed):
Thank you @joshuatf for addressing the changes! The `initialValues` issue was a pretty tricky thing to deal with.
Good job! :shipit:

octaedro reviewed (approved):
🚀
